src/looprestoration.rs

```diff
@@ -726,3 +726,223 @@ pub(crate) unsafe extern "C" fn selfguided_filter<BD: BitDepth>(
+        dst.as_mut_ptr(),
+        tmp.as_mut_ptr(),
+        390 as libc::c_int as ptrdiff_t,
+        w,
+        h,
+        9 as libc::c_int,
+        (*params).sgr.s1,
+        bd,
+    );
+    let w1 = (*params).sgr.w1 as libc::c_int;
+    let mut j = 0;
+    while j < h {
+        let mut i = 0;
+        while i < w {
+            let v = w1 * dst[(j * 384 + i) as usize].as_::<libc::c_int>();
+            *p.offset(i as isize) = iclip(
+                (*p.offset(i as isize)).as_::<libc::c_int>()
+                    + (v + ((1 as libc::c_int) << 10) >> 11),
+                0 as libc::c_int,
+                bd.bitdepth_max().as_(),
+            )
+            .as_();
+            i += 1;
+        }
+        p = p.offset(BD::pxstride(stride as usize) as isize);
+        j += 1;
+    }
+}
+
+// TODO(randomPoison): Temporarily pub until init logic is deduplicated.
+pub(crate) unsafe extern "C" fn sgr_mix_c<BD: BitDepth>(
+    mut p: *mut libc::c_void,
+    stride: ptrdiff_t,
+    left: *const libc::c_void,
+    mut lpf: *const libc::c_void,
+    w: libc::c_int,
+    h: libc::c_int,
+    params: *const LooprestorationParams,
+    edges: LrEdgeFlags,
+    bitdepth_max: libc::c_int,
+) {
+    sgr_mix_rust(
+        p.cast(),
+        stride,
+        left.cast(),
+        lpf.cast(),
+        w,
+        h,
+        params,
+        edges,
+        BD::from_c(bitdepth_max),
+    )
+}
+
+unsafe fn sgr_mix_rust<BD: BitDepth>(
+    mut p: *mut BD::Pixel,
+    stride: ptrdiff_t,
+    left: *const [BD::Pixel; 4],
+    mut lpf: *const BD::Pixel,
+    w: libc::c_int,
+    h: libc::c_int,
+    params: *const LooprestorationParams,
+    edges: LrEdgeFlags,
+    bd: BD,
+) {
+    let mut tmp: [BD::Pixel; 27300] = [0.as_(); 27300];
+    let mut dst0: [BD::Coef; 24576] = [0.as_(); 24576];
+    let mut dst1: [BD::Coef; 24576] = [0.as_(); 24576];
+    padding::<BD>(&mut tmp, p, stride, left, lpf, w, h, edges);
+    selfguided_filter(
+        dst0.as_mut_ptr(),
+        tmp.as_mut_ptr(),
+        390 as libc::c_int as ptrdiff_t,
+        w,
+        h,
+        25 as libc::c_int,
+        (*params).sgr.s0,
+        bd,
+    );
+    selfguided_filter(
+        dst1.as_mut_ptr(),
+        tmp.as_mut_ptr(),
+        390 as libc::c_int as ptrdiff_t,
+        w,
+        h,
+        9 as libc::c_int,
+        (*params).sgr.s1,
+        bd,
+    );
+    let w0 = (*params).sgr.w0 as libc::c_int;
+    let w1 = (*params).sgr.w1 as libc::c_int;
+    let mut j = 0;
+    while j < h {
+        let mut i = 0;
+        while i < w {
+            let v = w0 * dst0[(j * 384 + i) as usize].as_::<libc::c_int>()
+                + w1 * dst1[(j * 384 + i) as usize].as_::<libc::c_int>();
+            *p.offset(i as isize) = iclip(
+                (*p.offset(i as isize)).as_::<libc::c_int>()
+                    + (v + ((1 as libc::c_int) << 10) >> 11),
+                0 as libc::c_int,
+                bd.bitdepth_max().as_(),
+            )
+            .as_();
```

Comment:
```suggestion
            *p.offset(i as isize) = bd.iclip_pixel(
                (*p.offset(i as isize)).as_::<libc::c_int>()
                    + (v + ((1 as libc::c_int) << 10) >> 11),
            );
```

I think it should be this?  Double-check, but it needs to be `bd.iclip_pixel(...)`.

Reply:
Done.